`Development`: Add V2 Messages Endpoint

MichaelOwenDyer opened the pull request:
This PR introduces a version 2 of the messages endpoint, which returns all content generated in a guidance program as opposed to just the one variable titled 'response', and additionally accepts just a string as the template parameter instead of an object containing a string and an unused id.

MichaelOwenDyer commented:
Tried running the formatter but can't seem to get the formatting checks to pass

dakennguyen commented:
> Tried running the formatter but can't seem to get the formatting checks to pass

you can try to run this `poetry run black .`

MichaelOwenDyer commented:
You're right, it will break production. There is [another PR](https://github.com/ls1intum/Artemis/pull/7352) open on Artemis to change that side as well. But it might actually be smarter to just create a /v2/ endpoint, that way this can be merged without Artemis breaking.

dakennguyen commented on app/services/guidance_wrapper.py:
```diff
@@ -32,6 +31,14 @@ def query(self) -> Content:
+        # by the handlebars template
+        # This regex matches strings like: {{gen 'response' temperature=0.0 max_tokens=500}}
+        # and extracts the variable name 'response'
+        import re
```
one small thing is that I think we should import the library at the top of the file

https://peps.python.org/pep-0008/#imports

> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

Hialus commented:
> @MichaelOwenDyer I think the endpoint POST /api/v1/messages is currently being used. Will this PR break production? Otherwise, we need to introduce POST /api/v2/messages or make sure it backward backward-compatible.

> You're right, it will break production. There is https://github.com/ls1intum/Artemis/pull/7352 open on Artemis to change that side as well. But it might actually be smarter to just create a /v2/ endpoint, that way this can be merged without Artemis breaking.

That is a great idea. I think this would make sense here for the transition phase, until Artemis Prod is updated. Then we could simply the v1 afterwards.

Hialus commented on app/services/guidance_wrapper.py:
```diff
@@ -32,6 +31,14 @@ def query(self) -> Content:
+        # This regex matches strings like: {{gen 'response' temperature=0.0 max_tokens=500}}
+        # and extracts the variable name 'response'
+        import re
+        pattern = r'{{(?:gen|geneach|set) [\'"]([^\'"]+)[\'"]$}}'
```
There is a small issue I can see here:
If there is more than one space between e.g. `gen` and the name e.g. `response`(like this: `gen     response`), this is still valid syntax, but is not caught by this regex. Also is the `$` the correct thing here? According to what I found it should be used at the end of a line, which is not the case here, as it is followed by `}}`.

MichaelOwenDyer replied:
Thanks for pointing this out, addressed these issues

MichaelOwenDyer commented:
Thanks for the reviews guys. I addressed the issues you pointed out, V1 API should be behaving equivalent to before (although the logic for getting the response variable is now local to the endpoint and not directly in the guidance wrapper). I also extended the message test to test both v1 and v2.

Hialus reviewed (changes requested):
Very nice that you could change this to a v1/v2 split!
Code changes look good, but the linter job is currently failing. Please fix that.

MichaelOwenDyer commented:
You got it sir. Those dangerously long comments have been shortened.

dakennguyen reviewed (approved):
code LGTM!

Hialus reviewed (approved):
Change look good to me and since it has been deployed to the test server for some time without issues it also is sufficiently tested now.
Great job 🚀
